https://en.wikipedia.org/wiki/Divisor_function
https://math.stackexchange.com/questions/964267/proofs-of-identity-for-product-of-binomial-coefficients

In [1]:
from time import time
from functools import reduce

In [46]:
def isPrime(n):
    sieve = [False,False]+[True]*(n-1)
    for i in range(2,n+1):
        if sieve[i]:
            for j in range (2,n//i+1):
                sieve[i*j] = False
    return sieve

In [47]:
def primeFact(n):
    factors = [{} for i in range(n+1)]
    for p in range(2,n+1):
        if is_prime[p]:
            e = 1
            q = p
            while q < n:
                for j in range (1,n//q+1):
                    if p in factors[q*j]:
                        factors[q*j][p] += 1
                    else:
                        factors[q*j][p] = 1
                e+=1
                q = p**e
    return factors

In [33]:
def B_fact(n):
    v = [0,0] + [2*k-n-1 for k in range(2,n+1)] 
    for i in range(2,n+1):
        if (v[i] != 0) and (not is_prime[i]):
            for p in prime_fact[i]:
                v[p] += prime_fact[i][p]*v[i]
            v[i] = 0
    return v

In [34]:
def potSumMod(n,e,m):
    s = 1
    for i in range(e):
        s = (s*n + 1)%m
    return s
    
def divSumMod(n, m):
    v = B_fact(n)
    p = 1
    for i in range(2,n+1):
        if v[i] != 0:
            p = (p*potSumMod(i,v[i],m))%m
    return p

In [35]:
def run(f):
    def wrapper(args):
        St = time()
        solution = f(*args)
        Et = time() - St 
        print("for args = ", args, "found", solution, "in", Et, "secs")
    return wrapper

In [36]:
@run
def Euler650(n,m):
    
    global is_prime
    global prime_fact
    
    is_prime = isPrime(n)
    prime_fact = primeFact(n)
    
    solution = 0
    
    for i in range(1,n+1):
        solution = (solution + divSumMod(i,m))%m
    return solution

In [48]:
#n = 20000
n = 100
m = 10**9+7
args = [n,m]
Euler650(args)

for args =  [100, 1000000007] found 332792866 in 0.04687333106994629 secs


for args =  [100, 1000000007] found 332792866 in 0.02992081642150879 secs
for args =  [200, 1000000007] found 271664942 in 0.17253684997558594 secs
for args =  [1000, 1000000007] found 361160563 in 12.15048885345459 secs
for args =  [2000, 1000000007] found 939425731 in 77.52625465393066 secs